### PR TITLE
feat(handlers): configurable parallel fan-in aggregation policy (#313)

### PR DESCRIPTION
**Pull request by clintecker**
## Summary

Engine half of #313 (defect 1): the parallel fan-in aggregation is no longer hardwired to success-if-any. Unblocked by dippin-lang v0.39.0, which carries `params:` on `parallel` / `fan_in` nodes (dippin-lang#110).

- **New policy params** on parallel and fan_in nodes: `fan_in_policy: any | all | quorum` plus `quorum: <n>`. Param names match the upstream dippin-lang v0.39.0 fixtures.
- **Default stays `any`** (success-if-any) — pinned by `TestParallelHandlerDefaultPolicyIsAny` / `TestFanInHandlerDefaultPolicyIsAny`, so workflows that set no policy see zero behavior change.
- **One policy, two code paths**: `ParallelHandler.aggregateStatus` and `FanInHandler` both evaluate the same `fanInPolicy` (shared in `pipeline/handlers/fanin_policy.go`), so the policy can be declared on either node.
- **Fail fast on misconfiguration**: unknown policy or `quorum` without a positive `n` errors before any branch is dispatched (no silent fallback — per CLAUDE.md error rules).
- **Observability**: a non-default policy's evaluation (success count + failed branch IDs) lands in the `EventParallelCompleted` message and in the `fan_in.policy_detail` context key (written on success too, so a re-review loop can't leave a stale "failed" detail).
- **Routing**: a policy-failed parallel node returns `OutcomeFail` and routes via normal `ctx.outcome = fail` edges — no engine special case, checkpoint resume unaffected (the engine still doesn't know about parallel/fan-in).
- **build_product.dip**: `ReviewParallel` opts into `fan_in_policy: all`, closing the canonical masked-adversarial-review hole — a single failed reviewer now routes to `EscalateReview` (the fail edge shipped in PR #312). The `CheckReviewsComplete` / `ClearStaleReviews` guard from PR #326 **stays** as defense in depth (it catches a reviewer that "succeeds" without writing its report, which no aggregation policy can see).
- dippin-lang bumped to v0.39.0 (`go get`, `PinnedDippinVersion` kept in sync).

## Scope decisions (please read)

- **Policy set**: shipped `any` / `all` / `quorum:<n>`; no required-branches variant. `all` covers build_product's adversarial-reviewer case directly, and quorum covers the lossy-majority case. Required-branches can be added later if a workflow needs it.
- **Defect 2 of #313 (per-branch `fallback_target` runtime-inert) is NOT fixed here.** The policy makes branch failure observable and routable at the aggregating node, which subsumes the practical cases, but a per-branch `fallback_target` is still inert. `build_product_with_superspec.dip`'s `ReviewParallel` was deliberately NOT flipped — it has no fail edge, so `all` would dead-stop it; needs its own escalation routing first. I'll comment on #313 with what remains and keep it open.

## Verification

- `go build ./...` and `GOOS=darwin GOARCH=arm64 go build ./...` — clean
- `go test ./... -short` — all packages pass
- `go test -race -short ./pipeline/...` — pass
- `dippin doctor` — ask_and_execute A/95, build_product A/90, build_product_with_superspec A/100 (baselines held)

Part of #313 (engine defect 1); leaves #313 open for defect 2.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/2389-research/codesmith/tracker/pr/344"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1783718107&installation_id=131176874&pr_number=344&repository=2389-research%2Ftracker&return_to=https%3A%2F%2Fgithub.com%2F2389-research%2Ftracker%2Fpull%2F344&signature=b256852a0c84b425a7f5369fa52654e2c2ab378a1db7a168619106b9e264fa78"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * Configurable fan-in aggregation for parallel workflows: any (default), all, quorum; policy outcomes include human-readable policy detail and failed branch IDs; context key "fan_in.policy_detail" added for diagnostics.

* **Bug Fixes**
  * Hardened example cross-review flow so a single reviewer failure now triggers escalation and prevents masked/silent failures.

* **Tests**
  * Added unit tests covering fan-in policy behaviors and edge cases.

* **Chores**
  * Bumped dippin language requirement to v0.39.0.

### CODE REVIEW

File: pipeline/handlers/parallel.go [hunk 1]
```diff
@@ -75,13 +82,19 @@ func (h *ParallelHandler) Execute(ctx context.Context, node *pipeline.Node, pctx
 	}
 	pctx.Set("parallel.results", string(resultsJSON))
 
-	status := aggregateStatus(collected)
+	status, policyDetail := aggregateStatus(collected, policy)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/parallel.go:88]: Valid — fixed in 03be9e3. An unsatisfied non-default policy now suppresses the join suggestion, so edge selection can't fall through `selectBySuggested` to the fan-in and mask the failure. The default `any` policy keeps suggesting the join on all-fail (back-compat — existing workflows route that failure at the fan-in node downstream); both behaviors are pinned by `TestParallelHandlerPolicyFailureSuppressesJoinSuggestion` / `TestParallelHandlerDefaultPolicyAllFailStillSuggestsJoin`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,28 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- **Configurable parallel fan-in aggregation policy** (issue #313, defect 1).
+  Parallel and fan-in nodes accept a `params:` block (dippin-lang v0.39.0)
+  with `fan_in_policy: any | all | quorum` (plus `quorum: <n>`). The default
+  stays `any` (success-if-any, back-compat); `all` requires every branch to
+  succeed; `quorum` requires at least `<n>` successful branches. Both
+  aggregation code paths honor the policy — `ParallelHandler`'s
+  `aggregateStatus` and `FanInHandler` — so the policy can be declared on
+  either node. Unknown policies and `quorum` without a positive `n` fail
+  fast before any branch is dispatched. A policy-caused failure names the
+  policy and the failed branch IDs in the `EventParallelCompleted` message,
```
[the remark this answers is not in the document]
Comment (clintecker) [reply CHANGELOG.md]: Valid — fixed in 77307ab. The entry now states invalid policy config is a hard configuration error and scopes the pre-dispatch timing to parallel nodes (fan_in nodes validate when they execute).

File: pipeline/handlers/fanin_policy.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
+// ABOUTME: Configurable fan-in aggregation policy shared by the parallel and fan-in handlers (#313).
+// ABOUTME: Supports any (default, success-if-any), all, and quorum:<n> branch-success policies.
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/fanin_policy.go]: Valid — fixed in 77307ab. The ABOUTME now names the actual param shape (`fan_in_policy: quorum` + `quorum: <n>`).

File: examples/build_product.dip [hunk 4]
```diff
@@ -1305,6 +1305,8 @@ workflow BuildProduct
       printf 'cleared stale review reports\n'
 
   parallel ReviewParallel -> ReviewClaude, ReviewCodex, ReviewGemini
+    params:
+      fan_in_policy: all
```
[the remark this answers is not in the document]
Comment (clintecker) [reply examples/build_product.dip:1321]: Valid — fixed in f5c7b9c. The ClearStaleReviews rationale now reflects `fan_in_policy: all` (a failed reviewer routes to EscalateReview) and names the remaining reason the wipe matters: a reviewer that reports success without writing its file is invisible to any aggregation policy.

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,33 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  with `fan_in_policy: any | all | quorum` (plus `quorum: <n>`). The default
+  stays `any` (success-if-any, back-compat); `all` requires every branch to
+  succeed; `quorum` requires at least `<n>` successful branches. Both
+  aggregation code paths honor the policy — `ParallelHandler`'s
+  `aggregateStatus` and `FanInHandler` — so the policy can be declared on
+  either node. Unknown policies and `quorum` without a positive `n` are
+  hard configuration errors (on a parallel node they fail before any branch
+  is dispatched; on a fan_in node, when it executes). A policy-caused
+  failure names the
+  policy and the failed branch IDs in the `EventParallelCompleted` message,
```
[the remark this answers is not in the document]
Comment (clintecker) [reply CHANGELOG.md]: Fixed in f5c7b9c — rewrapped, the sentence no longer splits mid-phrase.

File: pipeline/handlers/parallel.go [hunk 6]
```diff
@@ -75,21 +82,35 @@ func (h *ParallelHandler) Execute(ctx context.Context, node *pipeline.Node, pctx
...
 		NodeID:    node.ID,
-		Message:   fmt.Sprintf("fan-in complete, aggregate status: %s", status),
+		Message:   msg,
 	})
 
 	outcome := pipeline.Outcome{
 		Status:        status,
 		Stats:         aggregateBranchStats(collected),
 		ChildOverride: aggregateChildOverrides(branchOverridesOut),
 	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/parallel.go:107]: Valid — fixed in 4fbc355. The parallel handler now writes `fan_in.policy_detail` for non-default policies on both success and failure (success too, so a later pass can't leave a stale "failed" detail), keeping the breadcrumb available when the policy failure routes away from the fan-in. Default `any` writes nothing — back-compat pinned by `TestParallelHandlerPolicyDetailInContext`.

File: pipeline/handlers/parallel.go [hunk 7]
```diff
@@ -54,6 +54,13 @@ func (h *ParallelHandler) Execute(ctx context.Context, node *pipeline.Node, pctx
 		return pipeline.Outcome{}, err
 	}
 
+	// Validate the aggregation policy before dispatching any branch — a
+	// misconfigured policy must fail fast, not after burning branch work.
+	policy, err := resolveFanInPolicy(node.ID, node.ParallelConfig())
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/parallel.go:62]: Valid — fixed in 4fbc355. `Execute()`'s doc comment now describes the `fan_in_policy` aggregation semantics (any/all/quorum) instead of the old hardwired success-if-any.

File: pipeline/handlers/parallel.go [hunk 8]
```diff
@@ -75,22 +85,46 @@ func (h *ParallelHandler) Execute(ctx context.Context, node *pipeline.Node, pctx
...
+	}
+	if policy.name != "any" {
+		// Record the policy evaluation in context here too — a policy failure
+		// suppresses the join suggestion, so the fan-in node (the other
+		// fan_in.policy_detail writer) may never run, and diagnose/audit would
+		// otherwise lose the structured breadcrumb (Copilot, PR #344). Written
+		// on success as well so a later pass can't leave a stale "failed"
+		// detail in context.
+		outcome.ContextUpdates["fan_in.policy_detail"] = policyDetail
 	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/parallel.go:128]: Valid — fixed in 19ec1c1. Added `pipeline.ContextKeyFanInPolicyDetail` to pipeline/context.go and switched both handlers (and the tests) to it.

File: pipeline/handlers/fanin.go [hunk 9]
```diff
@@ -39,11 +45,21 @@ func (h *FanInHandler) Execute(_ context.Context, node *pipeline.Node, pctx *pip
...
 	}
+	if policy.name != "any" {
+		// Record the policy evaluation in context so the audit trail /
+		// diagnose can explain the routing (#313). Written on success too —
+		// pipeline context persists across re-review loops, so a fail-only
+		// write would leave a stale "failed" detail after a later pass
+		// succeeds. Successful-branch context still merges above so
+		// downstream escalation gates can reference partial output.
+		merged["fan_in.policy_detail"] = policy.detail(successes, len(results), failed)
+	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/fanin.go:62]: Valid — same fix as the parallel.go thread: `pipeline.ContextKeyFanInPolicyDetail` in pipeline/context.go, used by both writers and the tests (19ec1c1).
